## 6.1. Datasets cleaning and combining

## Table of contents:
### 1. Import libraries
### 2. Datasets cleaning
### 2.1. greenhouse_gas_emissions
### 2.2. temperature_change
### 2.3. population
### 3. Export clean dataframes
### 4. Merge dataframes
### 5. Descriptive analysis for merged dataframe
##

### 1. Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 2. Datasets cleaning

### 2.1. greenhouse_gas_emission

### -- Import dataset

In [2]:
# Create path variable
path = r'C:\Users\marta\OneDrive\Documents\A6 Personal Project'

In [3]:
# Import dataset
df_emissions = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'greenhouse_gas_emissions.csv'))

In [4]:
# Check the output
df_emissions.head()

Domain Code                                             Domain  \
0         GPP  Emissions from pre and post agricultural produ...   
1         GPP  Emissions from pre and post agricultural produ...   
2         GPP  Emissions from pre and post agricultural produ...   
3         GPP  Emissions from pre and post agricultural produ...   
4         GPP  Emissions from pre and post agricultural produ...   

   Area Code (M49)         Area  Element Code          Element  Item Code  \
0                4  Afghanistan          7273  Emissions (CO2)       6504   
1                4  Afghanistan          7273  Emissions (CO2)       6504   
2                4  Afghanistan          7273  Emissions (CO2)       6504   
3                4  Afghanistan          7273  Emissions (CO2)       6504   
4                4  Afghanistan          7273  Emissions (CO2)       6504   

                        Item  Year Code  Year      Value  
0  Fertilizers Manufacturing       2002  2002  44.587578  
1  Fertilizers Manufacturing       2003  2003  31.322626  
2  Fertilizers Manufacturing       2004  2004  36.437832  
3  Fertilizers Manufacturing       2005  2005  29.126283  
4  Fertilizers Manufacturing       2006  2006  33.988554

In [5]:
# Check the dimensions
df_emissions.shape

(89020, 11)

### -- Drop and rename columns

In [6]:
# Remove irrelevant columns (not needed for the analysis)
df_emissions = df_emissions.drop(columns = ['Domain Code', 'Domain', 'Element Code', 'Item Code', 'Year Code'])

In [7]:
# Check the result
df_emissions.head()

Area Code (M49)         Area          Element                       Item  \
0                4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
1                4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
2                4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
3                4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
4                4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   

   Year      Value  
0  2002  44.587578  
1  2003  31.322626  
2  2004  36.437832  
3  2005  29.126283  
4  2006  33.988554

In [8]:
# Rename columns 
df_emissions.rename(columns = {'Area Code (M49)' : 'country_code', 'Area' : 'country', 'Element' : 'emission_gas', 'Year' : 'year', 'Value': 'emissions', 'Item' : 'agriculture_prod_stage'}, inplace = True)

In [9]:
# Check the result
df_emissions.head()

country_code      country     emission_gas     agriculture_prod_stage  \
0             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
1             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
2             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
3             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
4             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   

   year  emissions  
0  2002  44.587578  
1  2003  31.322626  
2  2004  36.437832  
3  2005  29.126283  
4  2006  33.988554

### -- Check data types and missing values

In [10]:
# Check column type
df_emissions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89020 entries, 0 to 89019
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country_code            89020 non-null  int64  
 1   country                 89020 non-null  object 
 2   emission_gas            89020 non-null  object 
 3   agriculture_prod_stage  89020 non-null  object 
 4   year                    89020 non-null  int64  
 5   emissions               89020 non-null  float64
dtypes: float64(1), int64(2), object(3)
memory usage: 4.1+ MB


In [11]:
# Check missing values
df_emissions.isnull().sum()

country_code              0
country                   0
emission_gas              0
agriculture_prod_stage    0
year                      0
emissions                 0
dtype: int64

#### No missing values were found.

### -- Check for mixed-type data

In [12]:
# Check for mixed-type columns
for col in df_emissions.columns.tolist():
  weird = (df_emissions[[col]].applymap(type) != df_emissions[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_emissions[weird]) > 0:
    print (col)

#### There are no columns with mixed-type data.

### -- Check for duplicate values

In [13]:
# Create a subset dataframe with only duplicated values
df_dups = df_emissions[df_emissions.duplicated()]

In [14]:
# Check the result
df_dups

Empty DataFrame
Columns: [country_code, country, emission_gas, agriculture_prod_stage, year, emissions]
Index: []

#### No duplicates were found.

### -- Descriptive statistics

In [15]:
# Descriptive statistics
df_emissions.describe()

country_code          year      emissions
count  89020.000000  89020.000000   89020.000000
mean     425.221220   2011.567850     883.104582
std      253.750388      5.750006    9622.406187
min        4.000000   2002.000000       0.000000
25%      204.000000   2007.000000       0.003017
50%      417.000000   2012.000000       0.137046
75%      643.000000   2017.000000      19.831088
max      894.000000   2021.000000  485989.998950

#### Max value for emissions seems to be very large. Such big difference in values exists because some countries produce more emissions than others during each agriculture stage. Also, some production activities create higher emissions of certain types of gases. 

In [16]:
# Function to display all records 
pd.set_option('display.max_rows', None)

In [17]:
# Check the frequency for 'emisisons'
df_emissions ['emissions'].value_counts(dropna = False).sort_index

<bound method Series.sort_index of emissions
0.000000         488
0.000001         486
0.000002         377
0.000005         277
0.000004         211
0.000007         199
0.000003         191
0.000006         183
0.000010         180
0.000009         144
0.000008         141
0.000011         127
0.000021         120
0.000015         114
0.000034         109
0.000014         100
0.000019          95
0.000022          95
0.000020          94
0.000013          92
0.000025          91
0.000024          90
0.000012          87
0.000032          85
0.000017          84
0.000016          82
0.000029          82
0.000023          81
0.000026          79
0.000033          72
0.000018          71
0.000027          71
0.000053          69
0.000035          69
0.000028          65
0.000030          62
0.000031          61
0.000037          57
0.000043          53
0.000050          53
0.000051          52
0.000045          51
0.000036          50
0.000114          50
0.000058          50
0.000055  

In [21]:
# Check the shape after cleaning
df_emissions.shape

(89020, 6)

### 2.2. temperature_change

### -- Import dataset

In [22]:
# Import dataset
df_temp = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'temperature_change.csv'))

In [23]:
# Check the output
df_temp.head()

Domain Code                      Domain  Area Code (M49)         Area  \
0          ET  Temperature change on land                4  Afghanistan   
1          ET  Temperature change on land                4  Afghanistan   
2          ET  Temperature change on land                4  Afghanistan   
3          ET  Temperature change on land                4  Afghanistan   
4          ET  Temperature change on land                4  Afghanistan   

   Element Code             Element  Months Code               Months  \
0          7271  Temperature change         7020  Meteorological year   
1          7271  Temperature change         7020  Meteorological year   
2          7271  Temperature change         7020  Meteorological year   
3          7271  Temperature change         7020  Meteorological year   
4          7271  Temperature change         7020  Meteorological year   

   Year Code  Year  Value  
0       2002  2002  1.365  
1       2003  2003  0.587  
2       2004  2004  1.373  
3       2005  2005  0.401  
4       2006  2006  1.720

In [24]:
# Check the shape
df_temp.shape

(4768, 11)

### -- Drop and rename columns

In [25]:
# Remove irrelevant columns (not needed for the analysis)
df_temp = df_temp.drop(columns = ['Domain Code', 'Domain', 'Element Code', 'Element', 'Months Code', 'Months', 'Year Code'])

In [26]:
# Check the output
df_temp.head(3)

Area Code (M49)         Area  Year  Value
0                4  Afghanistan  2002  1.365
1                4  Afghanistan  2003  0.587
2                4  Afghanistan  2004  1.373

In [27]:
# Rename columns 
df_temp.rename(columns = {'Area Code (M49)' : 'country_code', 'Area' : 'country', 'Year' : 'year', 'Value': 'temp_change'}, inplace = True)

In [28]:
# Check the result
df_temp.head(3)

country_code      country  year  temp_change
0             4  Afghanistan  2002        1.365
1             4  Afghanistan  2003        0.587
2             4  Afghanistan  2004        1.373

### -- Check columns type and missing values

In [29]:
# Check columns data types
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4768 entries, 0 to 4767
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country_code  4768 non-null   int64  
 1   country       4768 non-null   object 
 2   year          4768 non-null   int64  
 3   temp_change   4547 non-null   float64
dtypes: float64(1), int64(2), object(1)
memory usage: 149.1+ KB


In [30]:
# Check missing values
df_temp.isnull().sum()

country_code      0
country           0
year              0
temp_change     221
dtype: int64

#### There are 221 missing values in 'temp_change' column. Records will be removed, as it's <1% of the dataset.

In [31]:
# Create a dataframe excluding misisng values
df_temp_clean = df_temp[df_temp['temp_change'].isnull() == False]

In [32]:
# Check the result
df_temp_clean.shape

(4547, 4)

### -- Check for mixed-type data

In [33]:
# Check for mixed-type columns
for col in df_temp_clean.columns.tolist():
  weird = (df_temp_clean[[col]].applymap(type) != df_temp_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_temp_clean[weird]) > 0:
    print (col)

#### No columns with mixed-type data.

### -- Check for duplicate values

In [34]:
# Create a subset dataframe with only duplicated values
df_dups_temp = df_temp_clean[df_temp_clean.duplicated()]

In [35]:
# Check the result
df_dups_temp

Empty DataFrame
Columns: [country_code, country, year, temp_change]
Index: []

#### No duplicates found.

### -- Descriptive statistics

In [36]:
# Descriptive statistics
df_temp_clean.describe()

country_code         year  temp_change
count   4547.000000  4547.000000  4547.000000
mean     432.117000  2011.480537     1.066530
std      253.431714     5.757205     0.542466
min        4.000000  2002.000000    -0.622000
25%      214.000000  2006.500000     0.707000
50%      428.000000  2011.000000     1.005000
75%      659.000000  2016.000000     1.371500
max      894.000000  2021.000000     5.327000

#### All values look realistic.

### 2.3. population

### -- Import dataset

In [37]:
# Import dataset
df_population = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'population.csv'))

In [38]:
# Check the output
df_population.head()

Domain Code             Domain  Area Code (M49)         Area  Element Code  \
0          OA  Annual population                4  Afghanistan           511   
1          OA  Annual population                4  Afghanistan           511   
2          OA  Annual population                4  Afghanistan           511   
3          OA  Annual population                4  Afghanistan           511   
4          OA  Annual population                4  Afghanistan           511   

                         Element  Item Code                       Item  \
0  Total Population - Both sexes       3010  Population - Est. & Proj.   
1  Total Population - Both sexes       3010  Population - Est. & Proj.   
2  Total Population - Both sexes       3010  Population - Est. & Proj.   
3  Total Population - Both sexes       3010  Population - Est. & Proj.   
4  Total Population - Both sexes       3010  Population - Est. & Proj.   

   Year Code  Year      Value  
0       2002  2002  21000.256  
1       2003  2003  22645.130  
2       2004  2004  23553.551  
3       2005  2005  24411.191  
4       2006  2006  25442.944

In [39]:
# Check the shape
df_population.shape

(4650, 11)

### -- Drop and rename columns

In [40]:
# Remove irrelevant columns (not needed for the analysis)
df_population = df_population.drop(columns = ['Domain Code', 'Domain', 'Element Code', 'Element', 'Item Code', 'Item', 'Year Code'])

In [41]:
# Check the output
df_population.head(3)

Area Code (M49)         Area  Year      Value
0                4  Afghanistan  2002  21000.256
1                4  Afghanistan  2003  22645.130
2                4  Afghanistan  2004  23553.551

In [42]:
# Rename columns 
df_population.rename(columns = {'Area Code (M49)' : 'country_code', 'Area' : 'country', 'Year' : 'year', 'Value': 'total_population'}, inplace = True)

In [43]:
df_population.head(3)

country_code      country  year  total_population
0             4  Afghanistan  2002         21000.256
1             4  Afghanistan  2003         22645.130
2             4  Afghanistan  2004         23553.551

### -- Check column data type and missing values

In [44]:
# Check columns data types
df_population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4650 entries, 0 to 4649
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   country_code      4650 non-null   int64  
 1   country           4650 non-null   object 
 2   year              4650 non-null   int64  
 3   total_population  4650 non-null   float64
dtypes: float64(1), int64(2), object(1)
memory usage: 145.4+ KB


In [45]:
# Check missing values
df_population.isnull().sum()

country_code        0
country             0
year                0
total_population    0
dtype: int64

#### No missing values found.

### -- Check for mixed-type data

In [46]:
# Check for mixed-type columns
for col in df_population.columns.tolist():
  weird = (df_population[[col]].applymap(type) != df_population[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_population[weird]) > 0:
    print (col)

#### No mixed-type columns found.

### -- Check for duplicates

In [47]:
# Create a subset dataframe with only duplicated values
df_dups_pop = df_population[df_population.duplicated()]

In [48]:
# Check the result
df_dups_pop

Empty DataFrame
Columns: [country_code, country, year, total_population]
Index: []

#### No duplicates found.

### -- Descriptive statistics

In [49]:
# Descriptive statistics
df_population.describe()

country_code         year  total_population
count   4650.000000  4650.000000      4.650000e+03
mean     434.142151  2011.560215      3.659184e+04
std      251.525318     5.761736      1.544948e+05
min        4.000000  2002.000000      5.110000e-01
25%      218.000000  2007.000000      4.030897e+02
50%      434.000000  2012.000000      5.204960e+03
75%      652.000000  2017.000000      1.992503e+04
max      894.000000  2021.000000      1.457935e+06

#### Values look realistic. Population values are in thousands, so max value of 1,4 billion is possible.

### 3. Export clean dataframes

In [50]:
# Export df_emissions as a csv file
df_emissions.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'gas_emissions_clean.csv'), index = False)

In [51]:
# Export df_temp_clean as csv file
df_temp_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'temperature_change_clean.csv'), index = False)

In [52]:
# Export df_population as csv file
df_population.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'population_clean.csv'), index = False)

### 4. Merge dataframes

In [53]:
# Merge df_temp_clean into df_emissions
df_emissions_temp = df_emissions.merge(df_temp_clean, on = ['country_code', 'country', 'year'], how = 'inner')

In [54]:
# Check the result
df_emissions_temp.head()

country_code      country     emission_gas     agriculture_prod_stage  \
0             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
1             4  Afghanistan  Emissions (CO2)             Food Transport   
2             4  Afghanistan  Emissions (CH4)             Food Transport   
3             4  Afghanistan  Emissions (N2O)             Food Transport   
4             4  Afghanistan  Emissions (CO2)                Food Retail   

   year  emissions  temp_change  
0  2002  44.587578        1.365  
1  2002  36.860881        1.365  
2  2002   0.003147        1.365  
3  2002   0.002172        1.365  
4  2002   2.357904        1.365

In [55]:
# Check the shape
df_emissions_temp.shape

(85307, 7)

In [56]:
# Merge df_emissions_temp with df_population
df_emissions_temp_pop = df_emissions_temp.merge(df_population, on = ['country_code', 'country', 'year'], how = 'inner')

In [57]:
# Check the result
df_emissions_temp_pop.head()

country_code      country     emission_gas     agriculture_prod_stage  \
0             4  Afghanistan  Emissions (CO2)  Fertilizers Manufacturing   
1             4  Afghanistan  Emissions (CO2)             Food Transport   
2             4  Afghanistan  Emissions (CH4)             Food Transport   
3             4  Afghanistan  Emissions (N2O)             Food Transport   
4             4  Afghanistan  Emissions (CO2)                Food Retail   

   year  emissions  temp_change  total_population  
0  2002  44.587578        1.365         21000.256  
1  2002  36.860881        1.365         21000.256  
2  2002   0.003147        1.365         21000.256  
3  2002   0.002172        1.365         21000.256  
4  2002   2.357904        1.365         21000.256

In [58]:
# Check the shape
df_emissions_temp_pop.shape

(84960, 8)

In [59]:
# Export final merged dataframe
df_emissions_temp_pop.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'emissions_temp_population.pkl'))

### 5. Descriptive analysis for merged dataframe

In [60]:
# Descriptive analysis
df_emissions_temp_pop.describe()

country_code          year      emissions   temp_change  \
count  84960.000000  84960.000000   84960.000000  84960.000000   
mean     423.685440   2011.531733     924.271448      1.106345   
std      253.024442      5.747401    9847.718329      0.534788   
min        4.000000   2002.000000       0.000000     -0.505000   
25%      203.000000   2007.000000       0.003663      0.736000   
50%      414.000000   2012.000000       0.155964      1.041000   
75%      643.000000   2017.000000      22.442834      1.413000   
max      894.000000   2021.000000  485989.998950      3.691000   

       total_population  
count      8.496000e+04  
mean       4.889179e+04  
std        1.823632e+05  
min        5.110000e-01  
25%        1.983465e+03  
50%        8.046828e+03  
75%        2.822518e+04  
max        1.457935e+06

In [61]:
# Check the frequency for 'country'
df_emissions_temp_pop ['country'].value_counts(dropna = False).sort_index

<bound method Series.sort_index of country
Slovenia                                                520
United Kingdom of Great Britain and Northern Ireland    520
Georgia                                                 520
Japan                                                   520
Italy                                                   520
China, mainland                                         520
Israel                                                  520
Türkiye                                                 520
Colombia                                                520
Bulgaria                                                520
India                                                   520
Ukraine                                                 520
Brazil                                                  520
Croatia                                                 520
Bosnia and Herzegovina                                  520
Hungary                                                 5

In [62]:
# Frequency distribution of 'agriculture_prod_stage'
df_emissions_temp_pop['agriculture_prod_stage'].value_counts(dropna = False).sort_index

<bound method Series.sort_index of agriculture_prod_stage
Food Transport                12435
Food Household Consumption    11877
Food Retail                   11571
Pesticides Manufacturing      11493
Industrial Wastewater          8152
Food Packaging                 7632
Domestic Wastewater            7342
Food Processing                5877
Solid Food Waste               3627
Fertilizers Manufacturing      2600
Incineration                   2354
Name: count, dtype: int64>

In [64]:
# Frequency distribution of 'emission_gas'
df_emissions_temp_pop['emission_gas'].value_counts(dropna = False).sort_index

<bound method Series.sort_index of emission_gas
Emissions (CH4)    31986
Emissions (N2O)    28705
Emissions (CO2)    24269
Name: count, dtype: int64>